pkg/verificationcaching/verificationcaching.go

```diff
@@ -0,0 +1,80 @@
+package verificationcaching
+
+import (
+	"github.com/trufflesecurity/trufflehog/v3/pkg/cache"
+	"github.com/trufflesecurity/trufflehog/v3/pkg/context"
+	"github.com/trufflesecurity/trufflehog/v3/pkg/detectors"
+	"github.com/trufflesecurity/trufflehog/v3/pkg/pb/detectorspb"
+)
+
+// FromDataCached executes detection on chunk data in a way that uses a provided verification cache to deduplicate
+// verification requests when possible.
+//
+// If the provided cache is nil, this function simply returns the result of the provided detector's FromData method.
+//
+// If verify is false, this function returns the result of the provided detector's FromData method. In this case, the
+// cache is only updated if forceCacheUpdate is true.
+//
+// If verify is true, and forceCacheUpdate is false, this function first executes the provided detector's FromData
+// method with verification disabled. Then, the cache is queried for each result. If they are all present in the cache,
+// the cached values are returned. Otherwise, the provided detector's FromData method is invoked (again) with
+// verification enabled, and the results are used to update the cache before being returned.
+//
+// If verify is true, and forceCacheUpdate is also true, the provided detector's FromData method is invoked, and the
+// results are used to update the cache before being returned.
+func FromDataCached(
+	ctx context.Context,
+	verificationCache cache.Cache[*detectors.Result],
+	getCacheKey func(result *detectors.Result) string,
+	detector detectors.Detector,
+	verify bool,
+	forceCacheUpdate bool,
+	data []byte,
+) ([]detectors.Result, error) {
+
+	if verificationCache == nil {
+		return detector.FromData(ctx, verify, data)
+	}
+
+	if !forceCacheUpdate {
+		withoutRemoteVerification, err := detector.FromData(ctx, false, data)
+		if err != nil {
+			return nil, err
+		}
+
+		if !verify {
+			return withoutRemoteVerification, nil
+		}
+
+		isEverythingCached := false
+		for _, r := range withoutRemoteVerification {
+			if cacheHit, ok := verificationCache.Get(getCacheKey(&r)); ok {
+				r.CopyVerificationInfo(cacheHit)
+			} else {
+				isEverythingCached = false
+				break
```

Comment:
I was playing around converting this to [slices.ContainsFunc](https://pkg.go.dev/slices@go1.23.2#ContainsFunc), but then I realized maybe it's not optimal to `break` here? If we run into something that isn't yet cached, everything after it won't be updated from cache. Maybe that's fine though?

(also dunno what I think about the `Func` functions in `slices`; I guess the main use case is to not dupe logic in loops, but that doesn't apply here)

Reply:
Yeah, this is just kinda unfortunate. If we find something that's not cached, we **have** to rerun `FromData`, which will return all the results for this chunk, whether they're cached or not. So there's no reason to read from the cache. (Did I successfully explain that?)

Reply:
> So there's no reason to read from the cache. (Did I successfully explain that?)

Essentially, this is a stopgap solution to handle duplicate chunks (e.g., scanning orgs + members that have multiple forks of `kubernetes/kubernetes`.)

Proper caching on a granular level will require splitting `FromData` and `Verify` into discrete functions.
```go
// pseudo-code
cache := cache.New()

for _, chunk := range chunks {
    for _, d := range detectors {
        results := d.FromData(ctx, chunk)
        
        for result := range results {
            if val, ok := cache.Get(detector.Type, result.Raw); ok {
                result.Update(val)
            } else {
                detector.Verify(result)
                cache.Set(detector.Type, result.Raw)
            }
        }
    }
}
```